services/ros/ros-version.service.js

```diff
@@ -0,0 +1,111 @@
+import gql from 'graphql-tag'
+import Joi from 'joi'
+import { renderVersionBadge } from '../version.js'
+import { GithubAuthV4Service } from '../github/github-auth-service.js'
+import { NotFound } from '../index.js'
+import { parseReleaseVersionFromDistro } from './ros-version-helpers.js'
+
+const tagsSchema = Joi.object({
+  data: Joi.object({
+    repository: Joi.object({
+      refs: Joi.object({
+        edges: Joi.array()
+          .items({
+            node: Joi.object({
+              name: Joi.string().required(),
+            }).required(),
+          })
+          .required(),
+      }).required(),
+    }).required(),
+  }).required(),
+}).required()
+
+const contentSchema = Joi.object({
+  data: Joi.object({
+    repository: Joi.object({
+      object: Joi.object({
+        text: Joi.string().required(),
+      }).allow(null),
+    }).required(),
+  }).required(),
+}).required()
+
+export default class RosVersion extends GithubAuthV4Service {
+  static category = 'version'
+
+  static route = { base: 'ros/v', pattern: ':distro/:packageName' }
+
+  static examples = [
+    {
+      title: 'ROS Package Index',
+      namedParams: { distro: 'humble', packageName: 'vision_msgs' },
+      staticPreview: {
+        ...renderVersionBadge({ version: '4.0.0' }),
+        label: 'humble',
+      },
+    },
+  ]
+
+  async handle({ distro, packageName }) {
+    const tagsJson = await this._requestGraphql({
+      query: gql`
+        query ($refPrefix: String!) {
+          repository(owner: "ros", name: "rosdistro") {
+            refs(
+              refPrefix: $refPrefix
+              first: 30
+              orderBy: { field: TAG_COMMIT_DATE, direction: DESC }
+            ) {
+              edges {
+                node {
+                  name
+                }
+              }
+            }
+          }
+        }
+      `,
+      variables: { refPrefix: `refs/tags/${distro}/` },
+      schema: tagsSchema,
+    })
+
+    // Filter for tags that look like dates: humble/2022-06-10
+    const tags = tagsJson.data.repository.refs.edges
+      .map(edge => edge.node.name)
+      .filter(tag => /^\d+-\d+-\d+$/.test(tag))
+
+    const ref = tags[0] ? `refs/tags/${distro}/${tags[0]}` : 'refs/heads/master'
+    const prettyRef = tags[0] ? `${distro}/${tags[0]}` : 'master'
```

Comment:
Question: If the user specified a particular distro and we can't find any corresponding tag, does it actually make sense to fall back to `master` here, or should we really just throw an exception saying we can't find any tags matching the requested distro here?

Reply:
It seems like the tagging scheme is relatively new, and only the currently active (not end-of-life) distributions have tags for releases. Older ones (e.g. `eloquent`) that have not been published in the recent past don't have tags, but the master branch still contains all the version info of packages from the old distros (the distro is also part of the file path that we read). I guess it's less likely that people will care to display versions for old distros, but it seemed like a reasonable fallback to me if there is not a current version tag available